add ecs guide (#260)
PrefectHQ/prefect-aws: docs/ecs_guide.md

# Deploy an ECS Worker in AWS

> ```mermaid

Reviewer: This pushes all the content below the fold for me — I don't think it should necessarily be at the top. I'd create a dedicated section for it "Architectural diagram" or something.

graph TB

  subgraph ecs_cluster[ECS cluster]
    subgraph ecs_service[ECS service]
      td_worker[Worker task definition] --> |restarts| prefect_worker((Prefect worker))
    end
    prefect_worker -->|kicks off| ecs_task
    
    subgraph ecs_task[ECS task for each flow run]
      fr_task_definition[Flow run task definition]
    end
  end

  subgraph prefect_cloud[Prefect Cloud]
    subgraph prefect_workpool[ECS work pool]
      default_workqueue[Default work queue]
    end
  end

  flow_code[Flow Code] --> |prefect project pull step| ecs_task
  prefect_worker -->|polls| default_workqueue
  prefect_workpool -->|configures| fr_task_definition
```
## Why use ECS for flow execution?

ECS (Elastic Container Service) tasks are a good option for executing Prefect 2 flow runs for several reasons:

1. **Scalability**: ECS scales your infrastructure in response to demand, effectively managing Prefect flow runs. ECS automatically administers container distribution across multiple instances based on demand.

> 1. **Flexibility**: ECS lets you choose between AWS Fargate and Amazon EC2 for container operation. Fargate abstracts the underlying infrastructure, while EC2 can start faster and offers additonal control over instance management and configuration. 

Reviewer: ```suggestion
1. **Flexibility**: ECS lets you choose between AWS Fargate and Amazon EC2 for container operation. Fargate abstracts the underlying infrastructure, while EC2 has faster job start times and offers additional control over instance management and configuration. 
```

> 1. **AWS Integration**: Connect with other AWS services, such as AWS IAM and CloudWatch.

Reviewer: ```suggestion
1. **AWS Integration**: Easily connect with other AWS services, such as AWS IAM and CloudWatch.
```

1. **Containerization**: ECS supports Docker containers and offers managed execution. Containerization encourages reproducable deployments.

!!! tip "ECS tasks != Prefect tasks"
    An ECS Task is **not** the same thing as a Prefect task. ECS tasks are run as part of an ECS Cluster, they launch containers as defined in the ECS Task definition. An [*ECS task definition*](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definitions.html) is the blueprint for the ECS task. It describes which Docker containers to run and what you want to have happen inside these containers.

## ECS in Prefect

> The ECS task running the Prefect worker should be an [*ECS service*](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_services.html), given its long-running nature and need for auto-recovery in case of failure. ECS services seek to maintain the requested number of tasks. An ECS service automatically replaces any task that fails, which is ideal for managing a long-running process such as a Prefect worker.

Reviewer 2: Really solid distinction, I will get the verbiage updated

> The ECS task running the Prefect worker should be an [*ECS service*](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_services.html), given its long-running nature and need for auto-recovery in case of failure. ECS services seek to maintain the requested number of tasks. An ECS service automatically replaces any task that fails, which is ideal for managing a long-running process such as a Prefect worker.

Author: Added information on what is a worker

> An ECS task definition creates ECS tasks. ECS tasks launch containers as specified in the task definition until the number of containers is met. This setup is ideal for ephemeral processes such as a Prefect flow run. 

Reviewer: ```suggestion
An ECS task definition describes creation of an ECS task. ECS tasks launch containers as specified in the task definition until the number of containers is met. This setup is ideal for ephemeral processes such as a Prefect flow run. 
```

When the ECS worker finds a scheduled flow run in the ECS work pool it is polling, it spins up the specified infrastructure on AWS ECS. The worker knows to build an ECS task definition for each flow run based on the configuration specified in the work pool. 

If you specify a task definition [ARN (Amazon Resource Name)](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference-arns.html) in the work pool, the worker will use that ARN when spinning up the EC2 instance, rather than creating a new ECS Fargate container.

You can use either EC2 or Fargate as the capacity provider. Fargate simplifies initiation and provides containerization, but lengthens infrastructure setup time for each flow run. Using EC2 for the ECS cluster can reduce setup time. In this example, we will show how to ue Fargate.
<hr>

!!! tip

>     If you prefer IaC (Infrastructure as Code) check out this [terraform module](https://github.com/PrefectHQ/prefect-recipes/tree/main/devops/infrastructure-as-code/aws/tf-prefect2-ecs-worker) to provision an ECS cluster with a worker.

Reviewer: No need to use this abbreviation as you're unlikely to use it again here
```suggestion
    If you prefer infrastructure as code check out this [Terraform module](https://github.com/PrefectHQ/prefect-recipes/tree/main/devops/infrastructure-as-code/aws/tf-prefect2-ecs-worker) to provision an ECS cluster with a worker.
```

## Prerequisites
Before you begin, make sure you have:

- An AWS account with permissions to create ECS services and IAM roles.
- The AWS CLI installed on your local machine. You can [download it from the AWS website](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html).
- An [ECS Cluster](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/clusters.html) to host both the worker and the flow runs it submits.
- A [VPC](https://docs.aws.amazon.com/vpc/latest/userguide/what-is-amazon-vpc.html) configured for your ECS tasks. A VPC is a good idea if using EC2 and required if using Fargate. TK is this true?

## Step 1: Set Up an ECS work pool
Before setting up the worker, create a simple [work pool](https://docs.prefect.io/latest/concepts/work-pools/#work-pool-configuration) of type ECS for the worker to pull work from.

Create a work pool from the Prefect UI or CLI:

> `prefect work-pool create --type ecs my-ecs-pool`

Reviewer: Can't suggest this here — but you should always use ``` code fencing when it's not inline

Author: good catch changed

Configure the VPC and ECS Cluster for your Work Pool via the UI:
![VPC](img/VPC_UI.png)

Configuring custom fields is easiest from the UI.

![ECSCluster](img/ECSCluster_UI.png)

!!! Warning
    You need to have a VPC specified for your work pool if you are using AWS Fargate.
    
![Launch](img/LaunchType_UI.png)
<sub>Pictures of UI Cropped

Next, set up an Prefect ECS worker that will discover and pull work from this work pool.
## Step 2: Set up an ECS worker
Start a Prefect worker in your ECS cluster.

Create a Docker image for your Prefect worker, which you can build and push to the Amazon ECR registry. TK do you have to?

For example, a minimal Dockerfile for an ECS worker could look like:
```Dockerfile
FROM prefecthq/prefect:2-python3.10
RUN pip install s3fs prefect-aws

> ```

Reviewer: I'd just use our official image `2-latest` then set the command to install `prefect-aws` before starting the worker.

Author: This is not considered best practice for production and as this is a guide as opposed to docs we are inclined to provide examples of more production-ready patterns.

Author: Done, made this change

### Step 3: Create an IAM Role for the ECS task

> To create an IAM role for the ECS task using the AWS CLI, follow these steps: 

Author: Authenticating to what? The IAM role is by far the most recommended pattern from our users using ECS

1. **Create a Trust Policy**

    The Trust Policy will specify that ECS can assume the role. 
    
    Save this policy to a file, such as `ecs-trust-policy.json`:

>     ```json

Reviewer: We should also definitely link out to AWS documentation for doing this from the console

Author: Added a link

    ```bash

    aws iam attach-role-policy \
    --role-name ecsTaskExecutionRole \ 
    --policy-arn arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy
    ```

Remember to replace the `--role-name` and `--policy-arn` with the actual role name and policy Amazon Resource Name (ARN) you want to use.

Now, you have a role named `ecsTaskExecutionRole` that you can assign to your ECS tasks. This role has the necessary permissions to pull container images and publish logs to CloudWatch.

### Step 4: Create a task definition

Next, create an ECS task definition that specifies the Docker image for the Prefect worker, the resources it requires, and the command it should run. In this example our command to run is `prefect worker start --pool my-ecs-pool`. 

Here are the steps:

1. Create a JSON file with the following contents:

```json
{
  "family": "prefect-worker-task",
  "networkMode": "awsvpc",
  "taskRoleArn": "<your-ecs-task-role-arn>",
  "executionRoleArn": "<your-ecs-task-role-arn>",
  "containerDefinitions": [
    {
      "name": "prefect-worker",
      "image": "<your-ecr-image>",
      "cpu": 512,
      "memory": 1024,
      "essential": true,

>       "command": [
>         "prefect",
>         "worker",
>         "start",
>         "--pool",
>         "my-ecs-pool"
>       ],

Reviewer: re my earlier comment this could be (i'd test it though!)

```suggestion
      "command": [
        "bash",
        "-c",
        "pip install prefect-aws && prefect worker start --pool my-ecs-pool",
      ],
```

Author: I've put:
```json
      "command": [
        "pip",
        "install",
        "prefect-aws",
        "&&",
        "prefect",
        "worker",
        "start",
        "--pool",
        "my-ecs-pool"
      ],
```

Reviewer: You can't normally use `&&` in a infra command like that. You'd need to use `bash -c` first. Does this actually work?

2. **Flexibility**: ECS lets you choose between AWS Fargate and Amazon EC2 for container operation. Fargate abstracts the underlying infrastructure, while EC2 has faster job start times and offers additional control over instance management and configuration.
3. **AWS Integration**: Easily connect with other AWS services, such as AWS IAM and CloudWatch.
4. **Containerization**: ECS supports Docker containers and offers managed execution. Containerization encourages reproducible deployments.

## ECS Flow Run Execution

To set up Prefect flow execution in your cloud infrastructure, you need to understand two key Prefect concepts: 

- a [worker](https://docs.prefect.io/2.10.11/concepts/work-pools/#worker-overview) to poll Prefect's API (specifically that worker's [work queue](https://docs.prefect.io/2.10.11/concepts/work-pools/#work-queues)) 
- and the emphemeral [infrastructure](https://docs.prefect.io/2.10.11/concepts/infrastructure/) that this worker will spin up for each scheduled [flow run](https://docs.prefect.io/2.10.11/concepts/flows/#flow-runs). 

#### Archetecture Diagram
```mermaid
graph TB

  subgraph ecs_cluster[ECS cluster]
    subgraph ecs_service[ECS service]

>       td_worker[worker task definition] --> |defines| prefect_worker((Prefect worker))

Author: Made this change

    end
    prefect_worker -->|kicks off| ecs_task
    fr_task_definition[flow run task definition]

    subgraph ecs_task["ECS Task Execution <br>  (flow run infrastructure)"]
    style ecs_task text-align:center
      
      flow_run((Flow Run))

    end
    fr_task_definition -->|blueprint for| ecs_task
  end

  subgraph prefect_cloud[Prefect Cloud]
    subgraph prefect_workpool[ECS work pool]
      default_workqueue[Default work queue]
    end
  end

>   flow_code["Flow Code (e.g. GitHub)"] --> |prefect project pull step| ecs_task

Author: Made this change

> ## Why use ECS for flow execution?

Reviewer 3: It'd be good to have a top-level title for this page.

```suggestion
# ECS Worker Guide

## Why use ECS for flow run execution?
```

ECS (Elastic Container Service) tasks are a good option for executing Prefect 2 flow runs for several reasons:

1. **Scalability**: ECS scales your infrastructure in response to demand, effectively managing Prefect flow runs. ECS automatically administers container distribution across multiple instances based on demand.
2. **Flexibility**: ECS lets you choose between AWS Fargate and Amazon EC2 for container operation. Fargate abstracts the underlying infrastructure, while EC2 has faster job start times and offers additional control over instance management and configuration.
3. **AWS Integration**: Easily connect with other AWS services, such as AWS IAM and CloudWatch.
4. **Containerization**: ECS supports Docker containers and offers managed execution. Containerization encourages reproducable deployments.

## ECS Flow Run Execution

To set up Prefect flow execution in your cloud infrastructure, you need to understand two key Prefect concepts: 

> - a [worker](https://docs.prefect.io/2.10.11/concepts/work-pools/#worker-overview) to poll Prefect's API (specifically that worker's [work queue](https://docs.prefect.io/2.10.11/concepts/work-pools/#work-queues)) 

Reviewer 3: ```suggestion
- a [worker](https://docs.prefect.io/latest/concepts/work-pools/#worker-overview) to poll Prefect's API (specifically that worker's [work queue](https://docs.prefect.io/latest/concepts/work-pools/#work-queues)) 
```

- and the emphemeral [infrastructure](https://docs.prefect.io/2.10.11/concepts/infrastructure/) that will be spun up by this worker for each scheduled [flow run](https://docs.prefect.io/2.10.11/concepts/flows/#flow-runs). 

#### Archetecture Diagram
```mermaid
graph TB

  subgraph ecs_cluster[ECS cluster]
    subgraph ecs_service[ECS service]
      td_worker[worker task definition] --> |restarts| prefect_worker((Prefect worker))
    end
    prefect_worker -->|kicks off| ecs_task
    fr_task_definition[flow run task definition]

    subgraph ecs_task["ECS Task Execution <br>  (flow run infrastructure)"]
    style ecs_task text-align:center
      
      flow_run((Flow Run))

    end
    fr_task_definition -->|blueprint for| ecs_task
  end

  subgraph prefect_cloud[Prefect Cloud]
    subgraph prefect_workpool[ECS work pool]
      default_workqueue[Default work queue]
    end
  end

  flow_code["Flow Code (e.g. GitHub)"] --> |prefect project pull step| ecs_task
  prefect_worker -->|polls| default_workqueue
  prefect_workpool -->|configures| fr_task_definition
```
### ECS in Prefect Terms
!!! tip "ECS tasks != Prefect tasks"

>     An ECS Task is **not** the same thing as a [Prefect task](https://docs.prefect.io/2.10.11/concepts/tasks/#tasks-overview). 

Reviewer 3: ```suggestion
    An ECS Task is **not** the same thing as a [Prefect task](https://docs.prefect.io/latest/concepts/tasks/#tasks-overview). 
```

>     ECS tasks are run as part of an ECS Cluster, they launch containers as defined in the ECS Task definition. 

Reviewer 3: ```suggestion
    ECS tasks are groupings of containers that run within an ECS Cluster. An ECS task's behavior is determined by its task definition. 
```

    
    An [*ECS task definition*](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_definitions.html) is the blueprint for the ECS task. It describes which Docker containers to run and what you want to have happen inside these containers.

ECS Tasks are instances of a task definition. A Task Execution launches container(s) as defined in the task definition **until they are stopped or exit on their own**. This setup is ideal for ephemeral processes such as a Prefect flow run.

> The ECS task running the Prefect [worker](https://docs.prefect.io/2.10.11/concepts/work-pools/#worker-overview) should be an [**ECS Service**](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_services.html), given its long-running nature and need for **auto-recovery in case of failure**. An ECS service automatically replaces any task that fails, which is ideal for managing a long-running process such as a Prefect worker.

Reviewer 3: ```suggestion
The ECS task running the Prefect [worker](https://docs.prefect.io/latest/concepts/work-pools/#worker-overview) should be an [**ECS Service**](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_services.html), given its long-running nature and need for **auto-recovery in case of failure**. An ECS service automatically replaces any task that fails, which is ideal for managing a long-running process such as a Prefect worker.
```

> When a Prefect [flow](https://docs.prefect.io/latest/concepts/flows/) is scheduled to run, that information is sent to the Prefect work pool specified in the flow's [deployment](https://docs.prefect.io/latest/concepts/deployments). [Work pools](https://docs.prefect.io/latest/concepts/work-pools/?h=work#work-pool-overview) are typed according to the infrastructure the flow will run on. The ECS work pool type is used flows executed as ECS tasks. The Prefect [worker](https://docs.prefect.io/latest/concepts/work-pools/#worker-types) polling the work pool for work will have a matching ECS type.

Reviewer 3: ```suggestion
When a Prefect [flow](https://docs.prefect.io/latest/concepts/flows/) is scheduled to run, the run scheduled in the Prefect work pool specified in the flow's [deployment](https://docs.prefect.io/latest/concepts/deployments). [Work pools](https://docs.prefect.io/latest/concepts/work-pools/?h=work#work-pool-overview) are typed according to the infrastructure the flow will run on. Flow runs scheduled in an `ecs` typed work pool are executed as ECS tasks. Only Prefect ECS [workers](https://docs.prefect.io/latest/concepts/work-pools/#worker-types) can poll an `ecs` typed work pool.
```

> When the ECS worker finds a scheduled flow run in the ECS work pool it is polling, it spins up the specified infrastructure on AWS ECS. The worker knows to build an ECS task definition for each flow run based on the configuration specified in the work pool.

Reviewer 3: ```suggestion
When the ECS worker receives a scheduled flow run from the ECS work pool it is polling, it spins up the specified infrastructure on AWS ECS. The worker knows to build an ECS task definition for each flow run based on the configuration specified in the work pool.
```

You can use either EC2 or Fargate as the capacity provider. Fargate simplifies initiation, but lengthens infrastructure setup time for each flow run. Using EC2 for the ECS cluster can reduce setup time. In this example, we will show how to use Fargate.

<hr>

# AWS CLI Guide

!!! tip
    If you prefer infrastructure as code check out this [Terraform module](https://github.com/PrefectHQ/prefect-recipes/tree/main/devops/infrastructure-as-code/aws/tf-prefect2-ecs-worker) to provision an ECS cluster with a worker.
### Prerequisites

Before you begin, make sure you have:

- An AWS account with permissions to create ECS services and IAM roles.
- The AWS CLI installed on your local machine. You can [download it from the AWS website](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html).
- An [ECS Cluster](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/clusters.html) to host both the worker and the flow runs it submits.

> - A [VPC](https://docs.aws.amazon.com/vpc/latest/userguide/what-is-amazon-vpc.html) configured for your ECS tasks. A VPC is a good idea if using EC2 and required if using Fargate. TK is this true?

Reviewer 3: I believe you need a VPC either way, but the launch type you're using determines if you need to tell Prefect which VPC to use. It will probably require some research to be sure.
```suggestion
- A [VPC](https://docs.aws.amazon.com/vpc/latest/userguide/what-is-amazon-vpc.html) configured for your ECS tasks. A VPC is a good idea if using EC2 and required if using Fargate.
```

                "Principal": {
                    "Service": "ecs-tasks.amazonaws.com"
                },
                "Action": "sts:AssumeRole"
            }
        ]
    }
    ```

2. **Create the IAM Role**

    Use the `aws iam create-role` command to create the role:

    ```bash

    aws iam create-role \ 
    --role-name ecsTaskExecutionRole \
    --assume-role-policy-document file://ecs-trust-policy.json
    ```

3. **Attach the Policy to the Role**

    Amazon has a managed policy named `AmazonECSTaskExecutionRolePolicy` that grants the permissions necessary for ECS tasks. Attach this policy to your role:

    ```bash

    aws iam attach-role-policy \
    --role-name ecsTaskExecutionRole \ 
    --policy-arn arn:aws:iam::aws:policy/service-role/AmazonECSTaskExecutionRolePolicy
    ```

    Remember to replace the `--role-name` and `--policy-arn` with the actual role name and policy Amazon Resource Name (ARN) you want to use.

    Now, you have a role named `ecsTaskExecutionRole` that you can assign to your ECS tasks. This role has the necessary permissions to pull container images and publish logs to CloudWatch.

4. **Create a task definition**

> Next, create an ECS task definition that specifies the Docker image for the Prefect worker, the resources it requires, and the command it should run. In this example our command to run is `prefect worker start --pool my-ecs-pool`.

Reviewer 3: ```suggestion
Next, create an ECS task definition that specifies the Docker image for the Prefect worker, the resources it requires, and the command it should run. In this example, the command to start our worker is `prefect worker start --pool my-ecs-pool`.
```

Here are the steps:

1. Create a JSON file with the following contents:

```json
{
  "family": "prefect-worker-task",
  "networkMode": "awsvpc",
  "taskRoleArn": "<your-ecs-task-role-arn>",
  "executionRoleArn": "<your-ecs-task-role-arn>",
  "containerDefinitions": [
    {
      "name": "prefect-worker",
      "image": "prefecthq/prefect:2-latest",
      "cpu": 512,
      "memory": 1024,
      "essential": true,
      "command": [
        "pip",
        "install",
        "prefect-aws",
        "&&",
        "prefect",
        "worker",
        "start",
        "--pool",
        "my-ecs-pool"
      ],
      "environment": [
        {
          "name": "PREFECT_API_URL",
          "value": "https://api.prefect.cloud/api/accounts/<your-account-id>/workspaces/<your-workspace-id>"
        },
        {
          "name": "PREFECT_API_KEY",
          "value": "<your-api-key>"
        }
      ]
    }
  ]
}
```

> - Use `prefect config view` to see the values for `PREFECT_API_URL`.

Reviewer 3: ```suggestion
- Use `prefect config view` to view the `PREFECT_API_URL` for your current Prefect profile.
```

```

Replace `<your-ecs-cluster>` with the name of your ECS cluster, `<path-to-task-definition-file>` with the path to the JSON file you created in Step 2, `<your-subnet-ids>` with a comma-separated list of your VPC subnet IDs, and `<your-security-group-ids>` with a comma-separated list of your VPC security group IDs.

!!! tip "Sanity check"
    The work pool page in the Prefect UI allows you to check the health of your workers - make sure your new worker is live!

### Step 5: Pick up a flow run with your new worker!

1. Write a simple test flow in a repo of your choice:

    `my_flow.py`

    ```python
    from prefect import flow, get_run_logger

    @flow
    def my_flow():
        logger = get_run_logger()
        logger.info("Hello from ECS!!")

    if __name__ == "__main__":
        my_flow()
    ```

2. Configure a project [pull step](https://docs.prefect.io/latest/concepts/projects/#the-pull-section) to allow the worker to access your flow code at runtime.

    !!! Tip "Create a [Prefect Project](https://docs.prefect.io/latest/tutorials/projects/#initializing-a-project)."

>         Run the following command at the base of whichever repo contains your flow. (See our [Projects Tutorial](https://docs.prefect.io/2.10.12/tutorial/projects/#initializing-a-project) for other recipes with other pull step options.)  

Reviewer 3: ```suggestion
        Run the following command at the base of whichever repo contains your flow. (See our [Projects Tutorial](https://docs.prefect.io/latest/tutorial/projects/#initializing-a-project) for other recipes with other pull step options.)  
```

            repository: https://github.com/git-user/my-repo-name.git
            branch: main
            access_token: null
        ```

3. Deploy the flow to the server, specifying the ECS work pool

    ```bash
    prefect deploy my_flow.py:my_flow --name ecs-test-deployment --pool my-ecs-pool
    ```

4. Find the deployment in the UI and click the **Quick Run** button!

## Optional Next Steps

1. Now that you are confident your ECS worker is healthy, you can experiment with different work pool configurations.

    - Do your flow runs require higher `CPU`?
    - Would an EC2 `Launch Type` speed up your flow run execution?

    These infrastructure configuration values can be set on your ECS work pool or they can be [overriden on the deployment level](https://docs.prefect.io/latest/concepts/infrastructure/#kubernetesjob-overrides-and-customizations) if desired.

2. Create a custom image for your flow executions, which you can build and push to the Amazon ECR registry.

    For example, a minimal base Dockerfile for a flow run in ECS could look like:

    ```Dockerfile
    FROM prefecthq/prefect:2-python3.10
    RUN pip install s3fs prefect-aws
    ```

> 3. Consider adding a [build step](https://docs.prefect.io/2.10.12/concepts/projects/#the-build-section) to your Prefect Project `prefect.yaml` if you want to automatically build a Docker image and push it to the repository referenced by the image name each time `prefect deploy` is run.

Reviewer 3: ```suggestion
3. Consider adding a [build step](https://docs.prefect.io/latest/concepts/projects/#the-build-section) to your Prefect Project `prefect.yaml` if you want to automatically build a Docker image and push it to the repository referenced by the image name each time `prefect deploy` is run.
```

    ```yaml

>     build:
>     - prefect_docker.projects.steps.build_docker_image:
>     requires: prefect-docker>=0.2.0
>     image_name: my-repo/my-image
>     tag: my-tag
>     dockerfile: auto
>     push: true

Reviewer 3: ```suggestion
    build:
        - prefect_docker.projects.steps.build_docker_image:
            requires: prefect-docker>=0.2.0
            image_name: my-repo/my-image
            tag: my-tag
            dockerfile: auto
            push: true
```
